api/agent/timer_desired_state_feedback_test.go

```diff
@@ -0,0 +1,322 @@
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+package agent
+
+import (
+	"github.com/eclipse-kanto/update-manager/api/types"
+	"testing"
+	"time"
+
+	"github.com/eclipse-kanto/update-manager/test/mocks"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewDesiredStateFeedbackNotifier(t *testing.T) {
+	mockCtr := gomock.NewController(t)
+	defer mockCtr.Finish()
+
+	actions := []*types.Action{}
+	reportedActions := []*types.Action{}
+
+	t.Run("test_new_desired_state_feedback_notifier", func(t *testing.T) {
+		mockClient := mocks.NewMockUpdateAgentClient(mockCtr)
+		interval := time.Second
+		updAgent := &updateAgent{
+			client: mockClient,
+		}
+		expectedNotifier := &desiredStateFeedbackNotifier{
+			interval:        interval,
+			agent:           updAgent,
+			actions:         actions,
+			reportedActions: reportedActions,
+		}
+		actualNotifier := newDesiredStateFeedbackNotifier(interval, updAgent)
+
+		assert.Equal(t, expectedNotifier, actualNotifier)
+	})
+}
+
+func TestDesiredStateFeedbackTimerSet(t *testing.T) {
+	mockCtr := gomock.NewController(t)
+	defer mockCtr.Finish()
+
+	actions := []*types.Action{
+		{
+			Component: &types.Component{
+				ID:      "action 1 component",
+				Version: "1.0.0",
+			},
+		},
+	}
+
+	actions2 := []*types.Action{
+		{
+			Component: &types.Component{
+				ID:      "action2",
+				Version: "2.0.0",
+			},
+		},
+	}
+
+	reportedActions := []*types.Action{
+		{
+			Component: &types.Component{
+				ID:      "action 1 component",
+				Version: "1.0.0",
+			},
+		},
+	}
+	reportedActions2 := []*types.Action{
+		{
+			Component: &types.Component{
+				ID:      "action2",
+				Version: "2.0.0",
+			},
+		},
+	}
+
+	t.Run("test_set_internal_timer_not_nil", func(t *testing.T) {
+		mockClient := mocks.NewMockUpdateAgentClient(mockCtr)
+		interval := time.Second
+		updAgent := &updateAgent{
+			client: mockClient,
+		}
+
+		notifier := newDesiredStateFeedbackNotifier(interval, updAgent)
+		notifier.internalTimer = time.AfterFunc(interval, func() {})
+
+		notifier.set(testActivityID, actions)
+
+		assert.Equal(t, actions, notifier.actions)
+		if notifier.internalTimer != nil {
+			if !notifier.internalTimer.Stop() {
+				<-notifier.internalTimer.C
+			}
+		}
+	})
+
+	t.Run("test_set_internal_timer_nil", func(t *testing.T) {
+		mockClient := mocks.NewMockUpdateAgentClient(mockCtr)
+		interval := time.Second
+		updAgent := &updateAgent{
+			client: mockClient,
+		}
+
+		notifier := newDesiredStateFeedbackNotifier(interval, updAgent)
+		mockClient.EXPECT().SendDesiredStateFeedback(gomock.Any(), gomock.Any()).Times(1)
+		notifier.set(testActivityID, actions)
+
+		assert.Equal(t, reportedActions, notifier.reportedActions)
+		assert.Equal(t, testActivityID, notifier.activityID)
+		if notifier.internalTimer != nil {
```

Comment:
is there a chance where the `internalTimer` will be initialized during this test case? If that's not the case this check is obsolete